### PR TITLE
Multi-Agent RL Trainers [Part 2 of Multi-Agent RL]

### PR DESCRIPTION
**Pull request by ASzot**
## Motivation and Context

Adds population play multi-agent RL algorithm. Command to run:
```
python habitat_baselines/run.py --config-name=rearrange/rl_hierarchical.yaml habitat_baselines.evaluate=False habitat_baselines/rl/policy=hl_neural habitat_baselines/rl/policy/hierarchical_policy/defined_skills=oracle_skills_ma habitat_baselines/rl/agent=pop_play benchmark/rearrange=multi_agent_tidy_house.yaml
```

## How Has This Been Tested

Local testing. 

## Types of changes

- New feature (non-breaking change which adds functionality)

### CODE REVIEW

File: habitat-baselines/habitat_baselines/config/default_structured_configs.py [hunk 1]
```diff
@@ -347,6 +347,14 @@ class DDPPOConfig(HabitatBaselinesBaseConfig):
 @attr.s(auto_attribs=True, slots=True)
 class TrainerAgentConfig(HabitatBaselinesBaseConfig):
     name: str = "SingleAgentAccessMgr"
+    ###############################
+    # Population play configuration
```
Comment (vincentpierre) [inline habitat-baselines/habitat_baselines/config/default_structured_configs.py:351]: What is the default here? SingleAgentAccessMgr seems to indicate there is only one agent in the AgentAccessMgr, but the attributes bellow seems to indicate population based training is enabled.

File: habitat-baselines/habitat_baselines/config/habitat_baselines/rl/agent/pop_play.yaml [hunk 2]
```diff
@@ -0,0 +1 @@
+name: "MultiAgentAccessMgr"
```
Comment (vincentpierre) [inline habitat-baselines/habitat_baselines/config/habitat_baselines/rl/agent/pop_play.yaml:1]: This is where num_total_agents etc... should be changed.

File: habitat-baselines/habitat_baselines/config/habitat_baselines/rl/agent/self_play.yaml [hunk 3]
```diff
@@ -0,0 +1,4 @@
+name: "MultiAgentAccessMgr"
```
Comment (vincentpierre) [inline habitat-baselines/habitat_baselines/config/habitat_baselines/rl/agent/self_play.yaml:1]: I think agent as the name of the folder is going to be confusing. I would stick to agent_manager or multi_agent_setup instead. (agent is already used in habitat-lab config)

File: habitat-baselines/habitat_baselines/rl/multi_agent/multi_agent_access_mgr.py [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+                    config,
+                    agent_env_spec,
+                    is_distrib,
+                    device,
+                    use_resume_state,
+                    num_envs,
+                    percent_done_fn,
+                    lr_schedule_fn,
+                )
+            )
```
Comment (vincentpierre) [inline habitat-baselines/habitat_baselines/rl/multi_agent/multi_agent_access_mgr.py:90]: I think this whole thing should be a helper method.

File: habitat-baselines/habitat_baselines/rl/multi_agent/multi_agent_access_mgr.py [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+    def actor_critic(self):
+        return self._multi_policy
+
+    @property
+    def updater(self):
+        return self._multi_updater
+
+    @property
+    def policy_action_space(self):
+        return self._agents[0].policy_action_space
```
Comment (vincentpierre) [inline habitat-baselines/habitat_baselines/rl/multi_agent/multi_agent_access_mgr.py:248]: Why only of the first agent? Isn't it supposed to be combined ?

File: habitat-baselines/habitat_baselines/rl/multi_agent/utils.py [hunk 6]
```diff
@@ -0,0 +1,35 @@
+from typing import Any, Dict, Union
```
Comment (vincentpierre) [inline habitat-baselines/habitat_baselines/rl/multi_agent/utils.py:1]: Can you add docstrings in this file?

File: habitat-baselines/habitat_baselines/rl/multi_agent/utils.py [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+
+from habitat_baselines.common.tensor_dict import TensorDict
+
+
+def _get_agent_k(name: str, agent_s: str) -> str:
```
Comment (vincentpierre) [inline habitat-baselines/habitat_baselines/rl/multi_agent/utils.py:6]: remove_agent_prefix ?

File: habitat-baselines/habitat_baselines/rl/multi_agent/utils.py [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+        for k, v in names.items()
+        if agent_s in k or not k.startswith("agent_")
+    }
+    if was_td_dict:
+        return TensorDict(ret)
+    else:
+        return ret
+
+
+def get_agent_name(k, agent_i):
```
Comment (vincentpierre) [inline habitat-baselines/habitat_baselines/rl/multi_agent/utils.py:29]: add_agent_prefix ?

File: habitat-baselines/habitat_baselines/rl/multi_agent/utils.py [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+    return f"agent_{agent_i}_{k}"
+
+
+def add_agent_names(source_dict, dest_dict, agent_i):
```
Comment (vincentpierre) [inline habitat-baselines/habitat_baselines/rl/multi_agent/utils.py:33]: update_dict_with_agent_prefix ?

File: habitat-baselines/habitat_baselines/rl/multi_agent/utils.py [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+    for k, v in source_dict.items():
+        dest_dict[f"agent_{agent_i}_{k}"] = v
```
Comment (vincentpierre) [inline habitat-baselines/habitat_baselines/rl/multi_agent/utils.py:35]: use the method you defined above

File: habitat-baselines/habitat_baselines/rl/multi_agent/pop_play_wrappers.py [hunk 11]
```diff
@@ -0,0 +1,257 @@
+from collections import defaultdict
+from typing import Dict, List
+
```
Comment (vincentpierre) [inline habitat-baselines/habitat_baselines/rl/multi_agent/pop_play_wrappers.py:3]: add typing to the code

File: habitat-baselines/habitat_baselines/rl/multi_agent/pop_play_wrappers.py [hunk 12]
```diff
@@ -0,0 +1,257 @@
...
+class MultiPolicy(Policy):
+    """
+    Wraps a set of policies. Splits inputs and concatenates outputs.
+    """
+
+    def __init__(self):
+        self._active_policies = []
+
+    def set_active(self, active_policies):
```
Comment (vincentpierre) [inline habitat-baselines/habitat_baselines/rl/multi_agent/pop_play_wrappers.py:25]: policy, updater and storage all have this `set_active` which is not part of the inherited class. I would consider adding an interface for that.
Also, set_active takes as input a policy, which means the policies are tracked externally? Would it be simpler to hav set_active take an index as input  and let the multiPolicy track the policies?

**Review by vincentpierre — COMMENTED** [review]
Looks good. What kind of tests did you run ? 
Can you add :
 - a smoke test to make sure we won't break this later by mistake
 - a Readme to describe how to run population based training (like what you havev in the PR description)

File: habitat-baselines/habitat_baselines/config/default_structured_configs.py (hunk 1 above)
Comment (ASzot) [reply habitat-baselines/habitat_baselines/config/default_structured_configs.py:362]: I will make the default 1 to make this more clear. It is only enabled if the AgentAccessMgr type is properly set.

File: habitat-baselines/habitat_baselines/config/habitat_baselines/rl/agent/self_play.yaml (hunk 3 above)
Comment (ASzot) [reply habitat-baselines/habitat_baselines/config/habitat_baselines/rl/agent/self_play.yaml:1]: This is out of scope for this PR. This change was made in the Part 1 PR which was targeting main. I can create a PR for main with this fix.

File: habitat-baselines/habitat_baselines/rl/multi_agent/multi_agent_access_mgr.py (hunk 5 above)
Comment (ASzot) [reply habitat-baselines/habitat_baselines/rl/multi_agent/multi_agent_access_mgr.py:248]: Oops, that was not intentional. Fixed.

File: habitat-baselines/habitat_baselines/rl/multi_agent/pop_play_wrappers.py (hunk 12 above)
Comment (ASzot) [reply habitat-baselines/habitat_baselines/rl/multi_agent/pop_play_wrappers.py:25]: I think it is best to let the AgentAccessMgr track the policies. The wrappers are meant to be light. `set_active` is only used by these multi-agent wrappers and not single agent variants so I don't think it belongs in the base interfaces.

**Review by vincentpierre — APPROVED** [review]
I do not have more comments on this PR. It no longer is targeting main and mostly changes baselines code so I think we can merge it.
🚢 🇮🇹

**Pull request merged**